core/psh/nc.c

```diff
@@ -0,0 +1,337 @@
+	struct sockaddr_in sockaddr_in;
+	socklen_t addrlen_in = sizeof(struct sockaddr_in);
+
+	memset(&sockaddr_in, 0, addrlen_in);
+	if (socktype == SOCK_STREAM && listen(sfd, 0) < 0) {
+		fprintf(stderr, "nc: Fail to listen on a socket!\n");
+		return -1;
+	}
+
+	while (1) {
+		FD_ZERO(&fds);
+		/* Since not all platforms handle signals,
+	 	 * we need to be able to stop the program by
+	 	 * writing EOF to stdin.
+	 	 */
+		FD_SET(STDIN_FILENO, &fds);
+		FD_SET(sfd, &fds);
+
+		if (select(sfd + 1, &fds, NULL, NULL, NULL) <= 0) {
+			return -1;
+		}
+		else if (FD_ISSET(STDIN_FILENO, &fds)) {
+			while ((c = getchar()) != EOF && c != '\n') { }
+			if (c == EOF)
+				return -1;
+		}
+		else {
+			if (socktype == SOCK_STREAM) {
+				if ((cfd = accept(sfd, sockaddr, NULL)) < 0) {
+					fprintf(stderr, "nc: Fail to receive connection!\n");
+					return -1;
+				}
+				return cfd;
+			}
+			else if (socktype == SOCK_DGRAM) {
+				/* In UDP case we wait for a first msg,
+				 * save its originator's address and connect
+				 * it.
+				 */
+				memset(buf, 0, sizeof(buf));
+				*addrlen = sizeof(struct sockaddr_in6);
+				if ((c = recvfrom(sfd, buf, sizeof(buf), 0,
+				                  (struct sockaddr *) &sockaddr_in,
+				                  &addrlen_in)) < 0) {
+					fprintf(stderr, "nc: Can't receive msg!\n");
+					return -1;
+				}
+				write(STDOUT_FILENO, buf, c);
+
+				if ((connect(sfd, (struct sockaddr *) &sockaddr_in,
+				             addrlen_in)) < 0) {
+					fprintf(stderr, "nc: Fail to connect!\n");
+					return -1;
+				}
+
+				return sfd;
+			}
+			else {
+				return -1;
+			}
+		}
+	}
+}
+
+static void psh_nc_socktalk(int fd)
+{
+	fd_set fds;
+	int n;
+	char buf[64];
+
+	while (1) {
+		FD_ZERO(&fds);
+		FD_SET(STDIN_FILENO, &fds);
+		FD_SET(fd, &fds);
+
+		if (select(fd + 1, &fds, NULL, NULL, NULL) <= 0) {
+			return;
+		}
+
+		if (FD_ISSET(STDIN_FILENO, &fds)) {
+			if ((n = read(STDIN_FILENO, buf, sizeof(buf))) <= 0)
+				return;
+
+			if (send(fd, buf, n, 0) <= 0)
+				return;
+			memset(buf, 0, n);
+		}
+
+		if (FD_ISSET(fd, &fds)) {
+			if ((n = recv(fd, buf, sizeof(buf), 0)) <= 0)
+				return;
+			write(STDOUT_FILENO, buf, n);
+			memset(buf, 0, n);
+		}
+	}
+}
+
+int psh_nc(int argc, char *argv[])
+{
+	struct sockaddr_storage srcsockaddr, dstsockaddr;
+	int c, l = 0;
+	int socktype = SOCK_STREAM;
+	char *srcaddr = NULL, *srcport = NULL, *dstaddr = NULL, *dstport = NULL;
+	int af_family = AF_UNSPEC;
+	int fd, cfd;
+	socklen_t addrlen;
+
+	while ((c = getopt(argc, argv, "lu46s:p:")) != -1) {
+		switch (c) {
+		case 'l':
+			l = 1;
+			break;
+		case 'u':
+			socktype = SOCK_DGRAM;
+			break;
+		case '6':
+			af_family = AF_INET6;
+			break;
+		case '4':
+			af_family = AF_INET;
+			break;
+		case 's':
+			srcaddr = optarg;
+			break;
+		case 'p':
+			srcport = optarg;
+			break;
+		default:
+			fprintf(stderr, "nc: Unknown option!\n");
+			return -EINVAL;
+		}
+	}
+
+	argc -= optind;
+	argv += optind;
+
+	if (l) {
```

Comment:
What is this `if (1) {` for?
`#if 0` is preffered form for dead code storage ;)

Reply:
It's `if (l)`. `l` stands for `listen` mode. I was trying to follow our short variables coding style, but maybe it is too implicit.